### PR TITLE
Major improvements in code extensibility, reduced API usage, eliminate redundancy, etc.

### PR DESCRIPTION
**Pull request by GaurangTandon**
My main motive in this four days attempt was to first reduce the heavy API usage by introducing filters. As I got engrossed in the code further, I managed to fix issues related to code extensibility, redudancy, refactoring, caching, etc. So, this turned out to be a slightly larger update than expected.

This PR builds on the code directly from the latest branch (dev2.2.15). I have tested all the functions [that I changed](https://www.diffchecker.com/firA7ggZ). They all work, except for a few quirks that were present since before. 

These are the most important changes:

1. remove [synchronous ajax call](https://xhr.spec.whatwg.org/#synchronous-flag) - "Synchronous XMLHttpRequest outside of workers is in the process of being removed from the web platform as it has detrimental effects to the end user’s experience." under `getFromAPI`
2. throttled all observer calls - it is a given that `MutationObserver` fires dozens of times a second, but the user we're designing this app for won't even notice changes faster than 250ms. 
3. the `apiParameter` list in `common.info.json` was redundant, since none of the files are actually using the corresponding methods `apiParameter` or `metaApiParameter`. So, I removed that.
4. `parseCrossSiteLinks` now also works for bare links in comments
5. four new helper functions under `sox.helpers` - as they are frequently used in the code
6. redefined the constant `currentAPIParameter` using the helper method above

Apart from that, all the other changes fix bugs, or make the code extensible, faster, less redundant, etc. However, there are a few things I could not fix:

1. `downvotedPostsEditAlert` - I don't know where to begin refactoring this from. It has three API calls, and more than four hundred lines of code. I can't even comprehend how it is doing what it's supposed to do. Perhaps, you might want to refactor it later yourself. At least reduce its length, its huge :O 
2. What is the `$anchor` in `showQuestionStateInSuggestedEditReviewQueue`? I think you forgot to define that variable.
3. What is the need for Yahoo! YQL API in `standOutDupeCloseMigrated`? The SE API already provides us with `migrated_from` and `migrated_to` details.
4. How do you handle more than one dupe links to be clicked on the clickable stamp for `standOutDupeCloseMigrated`? It is a known fact that moderators can mark more than one dupe targets for a single question.
5. `hotNetworkQuestionsFiltering` is unable to hide those links which are under the "more hot questions" link. As soon as you click that link, all the list items under that automatically appear. I think it needs an onclick handler for that.
6. `standOutDupeCloseMigrated` dupe stamp link is `/q/ID` when it is set, but becomes `/undefined` when you click on it. 
7. the `getQuestionTags` part of `hotNetworkQuestionsFiltering` has been buggy since previous version. there is no next element sibling of `i.getQuestionTags`

----

Since this is a big change, feel free to go through it manually and comment and ping me for changes on individual lines - I'll be more than happy to explain my changes. Even though I have tested all the changed methods, feel free to test them again :)

And when you're happy, we can push these changes, possibly under a new `2.3.0` version number ;)

### CODE REVIEW

**Comment by shu8** [comment]
@GaurangTandon Wow, thanks a lot for your time! I will go through this as soon as I can, but I'm sure it'll be fine!

File: sox.features.js [hunk 1]
```diff
@@ -81,22 +81,19 @@
 
         copyCommentsLink: function() {
```
Comment (shu8) [inline sox.features.js]: Were changes to this function by accident? (It seems to have rolled back my changes)

File: sox.features.js [hunk 2]
```diff
@@ -685,14 +683,14 @@
...
 
             function addHotText() {
-                if (!$('.sox-hot').length) {
-                    $('#feed').html('<p>SOX: One of the 100 hot network questions!</p>');
+                if (!document.getElementsByClassName('sox-hot').length) {
+                    document.getElementById('feed').innerHTML = '<p>SOX: One of the 100 hot network questions!</p>';
 
                     //display:block to fix https://github.com/soscripted/sox/issues/243:
-                    $('#question-header').css('display', 'block').prepend('<div title="SOX: this is a hot network question!" ' + (sox.location.on('english.stackexchange.com') ? 'style="padding:13px"' : '') + ' class="sox-hot"><i class="fab fa-free-code-camp"></i><div>');
+                    $(document.getElementById('question-header')).css('display', 'block').prepend('<div title="SOX: this is a hot network question!" ' + (sox.location.on('english.stackexchange.com') ? 'style="padding:13px"' : '') + ' class="sox-hot"><i class="fab fa-free-code-camp"></i><div>');
```
Comment (shu8) [inline sox.features.js]: Is there any point in using vanilla JS just to make it a jQuery object straight after? :/

Comment (GaurangTandon) [reply sox.features.js]: Actually, `$(document.getElementById(ID))` is quite faster than `$("#ID")`...

File: sox.features.js (hunk 1 above)
Comment (GaurangTandon) [reply sox.features.js]: Probably. I do recall starting work on the latest dev branch. Ah...you went from `dev*.*.14` to `dev*.*.15` by the time I finished work, Hadn't noticed that. Sorry! Feel free to roll back to your original.

File: sox.features.js (hunk 2 above)
Comment (shu8) [reply sox.features.js]: @GaurangTandon fair enough, I never knew that!

**Comment by shu8** [comment]
@GaurangTandon merged! thanks again for all the changes -- it must have taken quite some time! :)

**Comment by GaurangTandon** [comment]
That's wonderful :D

**Comment by shu8** [comment]
@GaurangTandon I've fixed most of the issues you mentioned, but I couldn't reproduce the duplicate button's link being broken -- does it still happen? 

I have no idea why I used YQL for the migrated questions 😆 I also had something in there to check for 'migrated from' but AFAIK the migrated notice only shows on the 'from' site?

For multiple duplicates, it just shows one of them...

I'll be taking out downvotedPostsEditAlert soon.

**Comment by GaurangTandon** [comment]
If I go to [this search page](https://stackoverflow.com/search?q=duplicate%3A1), and click the first duplicate link, it immediately transforms from `/q/46898` into `/undefined`, taking me to `stackoverflow.com/undefined` :/  The link also changes if you right click on this link (verify in the dev tools).

I've little clue as to why this happens. If it possible that you have a rogue click event handler (probably a delegated one) somewhere in the code? :O 

I'm testing on a fresh Chrome profile specifically for testing SOX only, so effect of other addons is zero.

**Comment by shu8** [comment]
@GaurangTandon very weird :/ it only happens on the search page, not on any other page e.g. https://stackoverflow.com/questions. 

I'll look at features that act on the search page...

**Comment by shu8** [comment]
Fixed it!

---

If you're interested, it was:

     $(".js-search-results").on("mousedown touchstart", "div.result-link a", function (ev) {
         this.href = $(this).data("searchsession");
     });

on line 1332 of the /search page:

image

I honestly have no clue why it's there :/

I tried changing it to:

    var question = data.items[0],
          questionId = question.closed_details.original_questions[0].question_id,
          $linkToAdd = $('<a/>', {
              'style': 'display: inline',
              'href': 'https://' + sox.site.url + '/q/' + questionId,
              'click': function(e) {
                   e.preventDefault();
                   return;
               }
          }),
          $spanToAdd = $('<span>', {
              'class': 'standOutDupeCloseMigrated-duplicate',
              'title': 'click to visit duplicate',
              'html': '&nbsp;duplicate&nbsp;'
         });
         
         //styling for https://github.com/soscripted/sox/issues/181
         $anchor.after($linkToAdd.append($spanToAdd));

but that click handler never got triggered, no idea why.

So, in the end, I just added a `data-searchsession` attribute to the sox span itself, so the search code can work for us rather than against!

**Comment by GaurangTandon** [comment]
> I've fixed most of the issues you mentioned

This is really minor so I'll comment it here. `getQuestionTags` works but it immediately redirects to the linked question on click, because the `i.getQuestionTags` itself is part of the link.

**Comment by shu8** [comment]
I'm away for a few days but I'll look at that when I get back :)

**Comment by GaurangTandon** [comment]
Aye, this box itself, and also the `i.getQuestionTags` icon that is there before this box appears. Both should not be a part of the link.

**Comment by shu8** [comment]
@GaurangTandon should be fixed in [dev 2.2.22](https://github.com/soscripted/sox/raw/dev/sox.user.js) :)

**Comment by GaurangTandon** [comment]
Dude....that's probably an overkill :( Those `i.getQuestionTags` everywhere are getting into my eyes:

image

Look, the thing is that this `i.getQuestionTags` function isn't going to be used too often. So, we can simply set it to show the question's tags on hovering over the question link itself. Let me share a PR.

**Comment by shu8** [comment]
Fair enough, although I feel we need *some* visual indicator that you can hover over them, and add this to the feature's `extended_desc`. A simple 'hover to show tags' just under the 'Hot Network Questions' could do?

**Comment by GaurangTandon** [comment]
You _will_ hover over the links to click them anyway. Once you hover over them, the tags list immediately pops out below the question link! ;)

**Comment by shu8** [comment]
The idea is to see the question tags before you click. 

If you don't know to hover to show the tags, there's no reason you should hover (by clicking) at all because you just wouldn't go to the question.

**Comment by shu8** [comment]
I'm saying it's not an obvious feature, and there's no need for a user to accidentally realise they can show the tags by hovering when we can just add a line of text that makes it obvious!

**Comment by GaurangTandon** [comment]
Well, alright with that.

**Pull request merged**